source/guides/recipes/autoformatting.md

```diff
@@ -0,0 +1,55 @@
+(autoformatting)=
+# Autoformatting
+
+[`nixfmt`](https://github.com/NixOS/nixfmt) is the official Nix autoformatter.
+It is not yet wired up to any official tooling. Subscribe to
+[NixOS/nix PR #11252](https://github.com/NixOS/nix/pull/11252) for updates on that effort.
+
+Because `nixfmt` doesn't support formatting whole directory trees, you need
+additional tooling such as `treefmt`. The `nixfmt-tree` package provides a
+`treefmt` pre-configured to run `nixfmt` on all nix files in your project. Just
+add it to your shell:
+
+```nix
+mkShell {
+  packages = [ pkgs.nixfmt-tree ];
+}
+```
+
+Note: this assumes you're working in a git repository with a [`.git/index` file
+at the
+root](https://github.com/NixOS/nixpkgs/blob/bbcdf758d61b6095693d00dea8a75061bf89b264/pkgs/by-name/ni/nixfmt-tree/package.nix#L33-L34).
+
+If you need to change the root detection, or configure treefmt to format other (non-nix)
+files, you can use `treefmt.withConfig`:
+
+```nix
+pkgs.treefmt.withConfig {
+  runtimeInputs = [
+    pkgs.nixfmt-rfc-style
+    pkgs.ruff
+  ];
+
+  settings = {
+    # Customize detection of the root of the project.
+    tree-root-file = "flake.nix";
+
+    # Configure nixfmt for .nix files.
+    formatter.nixfmt = {
+      command = "nixfmt";
+      includes = [ "*.nix" ];
+    };
+
+    # And for .py file.
+    formatter.ruff = {
+      command = "ruff";
+      options = [ "format" ];
+      includes = [ "*.py" ];
+    };
+  };
+}
+```
+
+This can get a little tedious.
+[treefmt-nix](https://github.com/numtide/treefmt-nix) has a big library of
+preconfigured formatters, and provides a `check` derivation you can use in CI.
```

Comment:
If we're mentioning the `check` derivation, should we include a (simple) example of using it?

If flakes are acceptable in nix.dev guides, maybe adding the drv to a flake's `checks` output would be a suitable example?

Such an example may be a good opportunity to "sneak in" defining `formatter` and `devshell` outputs, although not if that ends up making the example less simple.

Reply:
I agree it would be nice to include a simple example of using the `check` derivation, but that requires either using npins or flakes to pull in treefmt-nix. It gets messy kind of fast, and I'm lightly leaning towards "just send them to treefmt-nix to figure this out", as I've done here.

I'm very ambivalent on this, though. I'm going to leave this unresolved, perhaps someone with strong opinions will show up.